### PR TITLE
[FAB-17628] Create RFC proposal for Config Transaction Library

### PR DESCRIPTION
**Pull request by caod123**

### CODE REVIEW

File: text/0000-config-transaction-library.md
Comment (denyeart) [inline text/0000-config-transaction-library.md:77]: Clarify whether the new repo will depend on Fabric for this package, or would the package just be copied into the new repo?

File: text/0000-config-transaction-library.md
Comment (denyeart) [inline text/0000-config-transaction-library.md]: This sentence was difficult for me to parse at first, probably because it jumps right in and mentions 'dependencies', while this library has not explicitly been called a 'dependency' above.

File: text/0000-config-transaction-library.md
Comment (denyeart) [inline text/0000-config-transaction-library.md]: Consumption by 'external users' makes it sound like end users could use it. I like how you discussed in the Epic the example of somebody using the library to build a CLI... saying something similar here would help people understand the intended usage of the library.

Comment (mastersingh24) [reply text/0000-config-transaction-library.md]:
```suggestion
clients for any operations involving channel config transactions. Currently `configtxgen` is available and frequently
```
not sure if `clients` or `applications` makes the most sense here

File: text/0000-config-transaction-library.md
Comment (mastersingh24) [inline text/0000-config-transaction-library.md]:
```suggestion
with the protolator tool and manually modifying fields prior to re-encoding to a protobuf and submitting for update.
```
or

```suggestion
with the protolator tool and manually modifying fields prior to encoding back to a protobuf and submitting for update.
```

File: text/0000-config-transaction-library.md
Comment (mastersingh24) [inline text/0000-config-transaction-library.md]:
```suggestion
Many potential consumers may already have a stable method for updating configuration transactions and may be
```

File: text/0000-config-transaction-library.md
Comment (mastersingh24) [inline text/0000-config-transaction-library.md:44]: The other potential drawback is that we are only going to provide the library in Golang

Comment (mastersingh24) [reply text/0000-config-transaction-library.md:44]: Of course I don't think this is a drawback as you can build CLIs or even a Restful / web application using Golang

File: text/0000-config-transaction-library.md
Comment (mastersingh24) [inline text/0000-config-transaction-library.md]:
```suggestion
Eventually it will become too onerous to maintain current workarounds for channel configuration management
```

File: text/0000-config-transaction-library.md
Comment (mastersingh24) [inline text/0000-config-transaction-library.md]:
```suggestion
Consumable Golang packages have consistent and effective APIs, are well-documented, leverage universal coding standards, and provide useful examples of common usage.
```

File: text/0000-config-transaction-library.md
Comment (mastersingh24) [inline text/0000-config-transaction-library.md]:
```suggestion
the Golang standard library and the following `hyperledger` packages and components:
```
nit but I think it's more correct to capitalize Golang

**Review by mastersingh24 — COMMENTED** [review]
I like the content overall.  Nice work.  Suggestions mostly focused on making things a bit more clear and some nit-picky writing style stuff

File: text/0000-config-transaction-library.md
Comment (caod123) [reply text/0000-config-transaction-library.md:77]: By dependencies this really just means imported packages. I'll add that to the sentence.

File: text/0000-config-transaction-library.md
Comment (caod123) [reply text/0000-config-transaction-library.md:44]: Noted, will mention that we may eventually support TinyGo compilation to web assembly since that also addresses that a bit.

**Comment by tock-ibm** [comment]
Will this library be able to generate a channel genesis block? Can it be used to bootstrap a network, or, in light of PR #24 to bootstrap an application channel?

**Comment by caod123** [comment]
> Will this library be able to generate a channel genesis block? Can it be used to bootstrap a network, or, in light of PR #24 to bootstrap an application channel?

Currently the library does not support creating a system channel create transaction, but the next track of work after the initial application channel work will focus on system channel. It was pushed back in priority because we're under the impression that system channel is currently in the works of being removed, but we would still like to support it for our API in the meantime.

The library will not be intended for bootstrapping a network and will only be used to build channel create transactions for application channels (and later system channels), as well as to modify config transactions and produce the desired config update envelope. There was some discussion around whether we'd like to eventually support something similar for peer configuration but that is far outside of the scope of this library.

There will be some considerations after the work in #24 is completed though, but for now we're creating the library as if #24 does not exist.

**Review by denyeart — APPROVED** [review]
LGTM now

**Comment by mastersingh24** [comment]
Thx @caod123 ... LGTM

**Review by mastersingh24 — APPROVED** [review]

File: text/0000-config-transaction-library.md
Comment (cendhu) [inline text/0000-config-transaction-library.md:77]: We also need to consider the imports made by the test files. Currently, the following are imported by the test 

- "github.com/hyperledger/fabric/common/tools/protolator"
- "github.com/hyperledger/fabric/common/tools/protolator/protoext/commonext"
- "github.com/hyperledger/fabric/common/tools/protolator/protoext/ordererext"
- "github.com/hyperledger/fabric/common/tools/protolator/protoext/peerext"

File: text/0000-config-transaction-library.md
Comment (cendhu) [inline text/0000-config-transaction-library.md]: For the following reasons, I think it's good to keep the `config` pkg in a separate repo. 

1. config pkg is a library used by _clients_ and neither by _peer_ nor _orderer_ (i assume this from the RFC)
2. the overall goal is to keep the fabric code minimal
3. we already have taken code out of fabric and created `fabric-proto-go` and `fabric-chaincode-go` repos. We do import them back into fabric repo. Even if we put `config` pkg in a separate repo, I assume we won't even import it into Fabric.

Due to these reasons, I think it might be good to have a separate repo. I am not sure whether `hyperledger/fabric-config` alone warrant a repo. Hence, we can also have a repo called `hyperledger/fabric-library` we can then add config library and other future libraries such as 
- `block_decoder` -- a library that can aid the client to open and inspect a block or a transaction
-  `snapshot` -- a library to process the snapshot file, which is created using ledger checkpoint, at the client side for creating a local database for analytics
- `peer-config` -- a library to create a peer configuration tx to dynamically change the internal peer configuration specified in the `core.yaml `without employing the current `stop-reconfigure-start` approach (assumption: we introduce a module at the peer to process local configuration tx to update these config values). 

I just speculate that these libraries might be needed in the longer-term when people want to build a lot of addon on top of Fabric. Hence, a folder called `channelconfig` under `hyperledger/fabric-library` might be a suitable one IMO.

File: text/0000-config-transaction-library.md
Comment (cendhu) [inline text/0000-config-transaction-library.md]: `config` is too generic IMO. We have `peer config`, `gossip config`, `ledger config`. The config values present in `core.yaml` are retrieved and these configs are constructed during the peer startup and passed to each initializer accordingly.

May be `channelconfig` or `configtx` can be used as it seems to match the description of the library.

> a consumable standalone library that supports generating configuration envelopes for operations including application and system channel creation, channel configuration updates, and attaching endorsing signatures to generated envelopes.

If there is some better name. it can be used as well but `config` looks like disinformation to me.

File: text/0000-config-transaction-library.md
Comment (cendhu) [inline text/0000-config-transaction-library.md]: Danny, the below comments are already discussed in the code review. I am just conveying my realization. 

_The client usage flow looks like a state-machine to me. Every API call can result in an `error` and every error result in a `panic`. The user cannot move from one state to another without passing certain criteria. Using these APIs, I am sure that the client may not submit any invalid channel config tx to orderer and it would remove the need for any code that validates the received tx at the orderer node (kidding, still we cannot remove it)._

 I think, for a general-purpose library such as golang `map`, `delete()` not returning error might be okay. After some thought (and some discussions with some maintainers), I realized that this is more of a special purpose library and hence, the `error` returns & state-machine approach are fine. 

We still need to think about whether we should return an `error` or [error type](https://golang.org/doc/effective_go.html#errors) to the library user.

File: text/0000-config-transaction-library.md
Comment (caod123) [reply text/0000-config-transaction-library.md:77]: I'll throw these all in under `fabric/common/tools/protolator` as the overarching component to save a couple lines

File: text/0000-config-transaction-library.md
Comment (caod123) [reply text/0000-config-transaction-library.md]: I agree, I'm open to the name of the repo and would be interested in hearing other's thoughts on where it should live whether that's `fabric-libraries` or some other name

File: text/0000-config-transaction-library.md
Comment (caod123) [reply text/0000-config-transaction-library.md]: `configtx` has already been decided upon as the name of the package. That rename will be part of moving it out of fabric but the initial package name is `pkg/config` hence why it's listed here as such.

File: text/0000-config-transaction-library.md
Comment (caod123) [reply text/0000-config-transaction-library.md]: Agreed, will have that conversation with the config squad as we wrap up on most of the feature work and move on to cleanup/improvements.

File: text/0000-config-transaction-library.md
Comment (caod123) [reply text/0000-config-transaction-library.md:77]: Although not explicitly mentioned as part of this RFC, do note that protolator will likely also move alongside the config package and live as a consumable library as well (used both by config package and fabric)

File: text/0000-config-transaction-library.md
Comment (denyeart) [inline text/0000-config-transaction-library.md]: the link should be 
https://godoc.org/github.com/hyperledger/fabric/pkg/configtx

Comment (jyellick) [inline text/0000-config-transaction-library.md]: I certainly understand the desire not to duplicate documentation, and to have to maintain it in two places, but a link to dynamically generated content, untied to any particular commit, version, etc. makes for a very poor historical document.

I realize this work was ongoing when we were still finalizing the RFC process, so we probably shouldn't view this one as a model moving forward anyway, but if the goal of an RFC is to agree on a path forward, pointing to "the current state of the code" doesn't exactly make sense.

Since this one is straddling the boundary between the RFC process, I'm fine as is, but I just wanted to go on record that linking to go-doc shouldn't be a model for future RFCs, and that instead, I would expect to see much more substantial details in the RFC itself.

**Review by jyellick — APPROVED** [review]

File: text/0000-config-transaction-library.md
Comment (mastersingh24) [reply text/0000-config-transaction-library.md]: We have https://github.com/hyperledger/fabric-lib-go which was for common libraries shred across Fabric projects (specifically fabric and fabric-ca).  I do think this should be it's own repository.   While it is possible to have multiple, independently versioned modules in the same repository, it is a bit tricky if not done right.

File: text/0000-config-transaction-library.md
Comment (caod123) [reply text/0000-config-transaction-library.md]: That makes a lot of sense, will note for future RFCs. Thanks!

File: text/0000-config-transaction-library.md
Comment (caod123) [reply text/0000-config-transaction-library.md]: I'm fine with `hyperledger/fabric-config` for now then

**Comment by denyeart** [comment]
No concerns raised during Final Comment Period, let's go ahead and merge.

File: text/0000-config-transaction-library.md
Comment (davidkhala) [inline text/0000-config-transaction-library.md:50]: Some years ago, I have similar ideas discussed with nodejs-sdk maintainers that we could make nodejs version of configtxgen or even configtxlator, It is always possible. I am quite happy that we could have this as an isolated go package. And if we plan to provide a Config Transaction Library on nodejs, I am one of volunteer.

**Pull request merged**
